1111-i386-txt-Add-Intel-TXT-ACM-module-support.patch

```diff
@@ -0,0 +1,612 @@
++ *     from this software without specific prior written permission.
++ *
++ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
++ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
++ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
++ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
++ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
++ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
++ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
++ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
++ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
++ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
++ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
++ * OF THE POSSIBILITY OF SUCH DAMAGE.
++ *
++ */
++
++/*
++ *  GRUB  --  GRand Unified Bootloader
++ *  Copyright (C) 2020  Oracle and/or its affiliates.
++ *
++ *  GRUB is free software: you can redistribute it and/or modify
++ *  it under the terms of the GNU General Public License as published by
++ *  the Free Software Foundation, either version 3 of the License, or
++ *  (at your option) any later version.
++ *
++ *  GRUB is distributed in the hope that it will be useful,
++ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
++ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
++ *  GNU General Public License for more details.
++ *
++ *  You should have received a copy of the GNU General Public License
++ *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
++ */
++
++#include <grub/loader.h>
++#include <grub/memory.h>
++#include <grub/normal.h>
++#include <grub/err.h>
++#include <grub/misc.h>
++#include <grub/types.h>
++#include <grub/safemath.h>
++#include <grub/dl.h>
++#include <grub/cpu/relocator.h>
++#include <grub/i386/cpuid.h>
++#include <grub/i386/msr.h>
++#include <grub/i386/txt.h>
++
++/*
++ * Returns hdr + offset if [offset, offset + count * size) is within the bounds
++ * of the ACM, NULL otherwise.
++ */
++static void*
++n_fit_in_acm (struct grub_txt_acm_header *hdr, grub_uint32_t offset,
++	      grub_uint32_t size, grub_uint32_t count)
++{
++  grub_uint32_t total_size, ret_u32, elem_end;
++  /* ACM size overflow was checked in is_acmod() */
++  grub_uint32_t acm_len = hdr->size * 4;
++
++  /* To simplify rest of the code check if offset is non-0 here */
++  if ( offset == 0 )
++    return NULL;
++
++  if ( grub_mul (size, count, &total_size) )
++    return NULL;
++
++  if ( grub_add (offset, total_size, &elem_end) )
++    return NULL;
++
++  if ( elem_end > acm_len )
++    return NULL;
++
++  if ( grub_add ((grub_uint32_t)(unsigned long)hdr, offset, &ret_u32) )
++    return NULL;
++
++  /* Not checking if (hdr + elem_end) overflows. We know that (hdr + acm_len)
++   * doesn't, and that elem_end <= acm_len. */
```

Comment:
Fair point, removed this check and updated comment accordingly.

As for overflow check in general, I agree that in most cases it shouldn't be necessary, but in this case there are no guaranties that each part lies within info_table entirely since there are many different versions of TXT in the use and only the newest version of specification is officially available. The changelog doesn't clearly state how the structures changed over the time, neither does their descriptions. In my opinion it is worth to do more checks than necessary than omit one, especially since this code is run once during booting, so performance isn't that important.